### PR TITLE
[SFN] base implementation of intrinsic functions

### PR DESCRIPTION
**Pull request by MEPalma**
Adds base: lexer, grammar, components, and implementation of stepfunctions' intrinsic functions.

Unrelated test failure:
`tests/integration/stepfunctions/legacy/test_stepfunctions_legacy.py::test_aws_sdk_task_delete_s3_object`

### CODE REVIEW

**Comment by whummer** [comment]
Let's review this PR once #7464 is merged and this PR has been rebased.. 👍

File: tests/integration/stepfunctions/v2/test_stepfunctions_v2.py [hunk 1]
```diff
@@ -404,10 +409,10 @@ def check_invocations():
         # clean up
         cleanup(sm_arn, state_machines_before, stepfunctions_client)
 
-    @pytest.mark.skip("Intrinsic Functions not yet supported.")
+    # @pytest.mark.skip("Intrinsic Functions not yet supported.")
```
Comment (dominikschubert) [inline tests/integration/stepfunctions/v2/test_stepfunctions_v2.py:412]: this should probably just be deleted right?

File: tests/integration/stepfunctions/v2/test_stepfunctions_v2.py [hunk 2]
```diff
@@ -404,10 +409,10 @@ def check_invocations():
...
     def test_intrinsic_functions(self, stepfunctions_client):
-        if os.environ.get("AWS_DEFAULT_REGION") != "us-east-1":
-            pytest.skip("skipping non us-east-1 temporarily")
+        # if os.environ.get("AWS_DEFAULT_REGION") != "us-east-1":
+        #     pytest.skip("skipping non us-east-1 temporarily")
```
Comment (dominikschubert) [inline tests/integration/stepfunctions/v2/test_stepfunctions_v2.py:415]: same here

**Review by dominikschubert — APPROVED** [review]
LGTM :+1: 

A few remarks:

- We can probably remove the number of files a bit by aggregating them which would be a bit more pythonic in general. E.g. Keep all the different function_argument_x classes in `function_argument.py`. This generally also makes reviewing, code navigation and refactoring in general a bit easier

- Do you think we should also verify the input argument types more? I'm not sure how stepfunction behaves when you for example try to add or multiply a number + string or when you have an array of non-string values. :thinking:

**Pull request merged**